Improve find-patient component

jniles opened the pull request:
This PR improves the find-patient component in a number of ways:
1. Uses an `angular.component()` instead of `angular.directive` for the find-patient component.
2. Renames `find-patient` to `bh-find-patient`. 
3. Optionally takes in a `required` parameter.
4. Adds e2e test hooks for `bhFindPatient` to the `shared/components.js` test file.
5. Uses `tabindex = -1` to ensure proper tab navigation
6. Uses [angular-moment](https://github.com/urish/angular-moment) to calculate a patient's age.
7. Adds the `PatientService.search()` method for other APIs to consume when trying to search for patients.  The `bhFindPatient` component uses this service to search for patients from the database.
8.  Ensures that the HTML attributes `autocomplete`, `autocapitalize`, etc are all turned off.

**Updates**
1.  Ensures the the "enter" key will submit the `<input>` element for find by id.
2. Nicer syntax for loading the saved options from the cache.

Addresses many of the issues in #69.

jniles committed:
chore(find-patient): migrate to component usage

This commit migrates to the find patient directive to a find-patient
component and renames it to `<bh-find-patient>`.  The functionality
remains identical, but the codebase has been drastically reduced.

Partially fixes #69 by using tabindex=-1 on non-important buttons.

jniles committed:
chore(find-patient): HTML improvements

This commit completes the HTML improvements suggested in #69 and starts
on the test harness for the find patient directive.

jniles committed:
chore(fp): rename findPatient and test harness

This commit renames the findPatient component to bhFindPatient to better
namespace it.  It also implements a few test hooks for the bhFindPatient
component using the following API:
 1. components.findPatient.findById(id)
 2. components.findPatient.findByName(name)

sfount commented on client/test/e2e/shared/components.js:
```diff
@@ -149,3 +149,63 @@ exports.currencyInput = {
 };
+
+/**
+ * hooks for the find patient component described in the component
```
:+1:

sfount commented on client/src/js/directives/bhFindPatient.js:
```diff
@@ -0,0 +1,263 @@
+  var LIMIT = 20;
+
+  /** supported searches: by name or by id */
+  vm.options = [{
```
I know this isn't your code (so I'm not sure where to post this) but this options configuration is actually quite difficult to follow - it seems to me that in a lot of places throughout this code we are relying on someone reading `selected.key === 0` as 'find by ID state' and `selected.key===1` as 'find by name state'. This leads to a lot of constants being used throughout both the controller and the template. 

Potentially as a configuration object is being defined anyway would it be possible to do something like 

``` js
options = { 
  findById : { ... } // (label, placeholder etc.)
  findByName : { ... } // (label, placeholder etc.)
};
```

This might then read

``` js
if (vm.selected === options.findById)
if (vm.selected === options.findByName)
```

jniles replied:
Sounds like a good improvement and thanks for the feedback.  This PR is probably the appropriate place to implement this change.  I'll put in another commit with this change.

jniles replied:
So...  You can implement this, but it makes `loadDefaultOption()` look ugly.  Here it is:

``` js
  function loadDefaultOption(option) {

   // default to findById
    if (!option) { findBy(vm.options.findById); }

    var byId = angular.equals(option, vm.options.findById);
    findBy(byId ? vm.options.findById : vm.options.findByName);
}
```

Granted, this is only called once.  Which do you prefer?

jniles committed:
chore(fp): rename template, syntax improvements

This commit contains three important changes:
 1. The template is now called bhFindPatient.tmpl.html for parity with
 the controller name.
 2. Fix: The "enter" key on the ID field now actually submits the <input>.
 This bug was caused by the HTML mistakenly thinking the <button> on the
 dropdown was a submit button and the enter key would navigate the
 selection to that button.  It is fixed with type="button".
 3. The patientList.tmpl.html now uses patient labels instead of a
 string concatenation, limiting the number of operations necessary.

jniles committed:
chore(fp): remove the submit button for typeahead

This commit removes the typeahead submit button to make the interface
less confusing and more uniform with other ways we've written inline
forms.

With this commit, all the improvements suggested to the find-patient
component so far have been implemented.

Closes #69.

sfount commented on client/src/partials/templates/bhFindPatient.tmpl.html:
```diff
@@ -0,0 +1,243 @@
+
+          <!-- search by name  -->
+          <input ng-model="$ctrl.nameInput"
+            ng-if="$ctrl.selected == $ctrl.options.findByName"
```
This is much more readable. Excellent.
